Add cross-compilation as sanity check

jossemargt-3pillar opened the pull request:
K3s downstream dependencies have support for Windows agents, making the need for windows specific minimal verifications. 

#### Proposed Changes ####

Add an extra CI cross-compiling K3s as a minimal verification. 

#### Types of Changes ####

<!-- What types of changes does your code introduce to K3s? Bugfix, New Feature, Breaking Change, etc -->
Build and CI

#### Verification ####

Successfully cross-compile K3s for windows x86_64 using `scripts/validate-cross-compilation`

#### Linked Issues ####

<!-- Link any related issues, pull-requests, or commit hashes that are relevant to this pull request. If you are opening a PR without a corresponding issue please consider creating one first, at https://github.com/k3s-io/k3s/issues . A functional example will greatly help QA with verifying/reproducing a bug or testing new features. -->

#5243 

#### User-Facing Change ####
<!--
Does this PR introduce a user-facing change? If no, just write "NONE" in the release-note block below.
If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

#### Further Comments ####

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

I have a couple questions that I'd like to address before moving on:

- I added build-windows as a separate drone step in the `amd64` pipeline. Is this the way we would tackle it or should I add it into the `scripts/ci` list? If so, when do you prefer build-windows to be executed after `build` or `package` (so we don't  package it without noticing it)?
- Based on RKE2 windows build I removed all the linker flags for static linking. Is this alright with the team? Otherwise we would need to install the required headers for `x86_64-w64-mingw32`

briandowns commented:
@luthermonson @phillipsj @rosskirkpat

brandond commented on scripts/build-windows:
```diff
@@ -0,0 +1,75 @@
+    -w -s"
+
+STATIC="-extldflags '-static'"
+# PR: Do we need the following linker flags for Windows binary?
+# STATIC="
+#     -extldflags '-static -lm -ldl -lz -lpthread'
```
IIRC these are all required to build a static binary, but since we're not shipping the resulting binary, just making sure that it builds, dropping these is probably fine.

jossemargt-3pillar replied:
I agree wit you. BTW I downloaded and grep'ed RKE2 windows build and there isn't any of these linker flags there.

```sh
grep -B5 -A10 "bin/rke2.exe -ldflags" [a temp dir]/logs_rancher_rke2_3007_1_2.log
+ echo
+ GOOS=windows
+ CGO_ENABLED=1
+ CXX=x86_64-w64-mingw32-g++
+ CC=x86_64-w64-mingw32-gcc
+ go build -tags ' netgo osusergo no_stage static_build sqlite_omit_load_extension ' -o bin/rke2.exe -ldflags '  -w 
        -X github.com/k3s-io/k3s/pkg/version.GitCommit=342347537760e72cef347ddb2dcbdcbb35754e30
        -X github.com/k3s-io/k3s/pkg/version.Program=rke2
        -X github.com/k3s-io/k3s/pkg/version.Version=v1.23.4-dev+e1aeae66
        -X github.com/rancher/rke2/pkg/images.DefaultRegistry=docker.io
        -X github.com/rancher/rke2/pkg/images.DefaultEtcdImage=rancher/hardened-etcd:v3.5.1-k3s1-build20220112
        -X github.com/rancher/rke2/pkg/images.DefaultKubernetesImage=rancher/hardened-kubernetes:v1.23.4-rke2r1-build20220217
        -X github.com/rancher/rke2/pkg/images.DefaultPauseImage=rancher/pause:3.6
        -X github.com/rancher/rke2/pkg/images.DefaultRuntimeImage=rancher/rke2-runtime:v1.23.4-dev-e1aeae66
'
```

brandond replied:
RKE2 on Windows doesn't have the same static/hardened requirements as the Linux builds. So yeah, I think it's fine.

phillipsj commented:
I don't think it will be an issue. Only concern is that isn't how it is consumed downstream which may uncover issues later.

jossemargt-3pillar commented:
Following up:

> - I added build-windows as a separate drone step in the amd64 pipeline. Is this the way we would tackle it or should I add it into the scripts/ci list? If so, when do you prefer build-windows to be executed after build or package (so we don't package it without noticing it)?
> - Based on RKE2 windows build I removed all the linker flags for static linking. Is this alright with the team? Otherwise we would need to install the required headers for x86_64-w64-mingw32

- After discussing with @galal-hussein . Either approach is fine, so I'll keep the validation as an independent CI step. Also, I renamed both the script and the step to not give incorrect expectations over its purpose.
- Since the resulting binary of this validation is not intended to be shipped, it is ok to not be statically compiled
